src/pybind/mgr/dashboard/services/nvmeof_conf.py

```diff
@@ -55,13 +55,19 @@ def add_gateway(cls, name, service_url, group, daemon_name):
         config = cls.get_gateways_config()
 
         if name in config.get('gateways', {}):
-            existing_gateways = config['gateways'][name]
-            for gateway in existing_gateways:
-                if 'daemon_name' not in gateway:
-                    gateway['daemon_name'] = daemon_name
-                    break
-                if gateway['service_url'] == service_url:
-                    return
+            # the nvmeof dashboard config used in v19.2.0 saves the below
+            # to a dict. Converting that to a list so that the upgrade
+            # properly migrate it to the newer format, and also keep it empty.
+            if isinstance(config['gateways'][name], dict):
+                config['gateways'][name] = []
+            else:
+                existing_gateways = config['gateways'][name]
+                for gateway in existing_gateways:
+                    if 'daemon_name' not in gateway:
+                        gateway['daemon_name'] = daemon_name
+                        break
+                    if gateway['service_url'] == service_url:
+                        return
```

Comment:
Regarding this piece of code, what's it's pourpose?:

```
    if 'daemon_name' not in gateway:
                        gateway['daemon_name'] = daemon_name
                        break
```

I understand that if it's missing the `daemon_name` information we are setting that, but then `break` happens and the function continues. Is that desired?

Won't the following code:
```
 new_gateway = {
            'service_url': service_url,
            'group': group,
            'daemon_name': daemon_name
        }

        if name in config.get('gateways', {}):
            config['gateways'][name].append(new_gateway)
        else:
            config['gateways'][name] = [new_gateway]
```
rewrite the gateway information?

Reply:
I think I added it for a backward compatibility a while ago. It might be irrelavant now but I didn't wanted to take risks by modifying it. Initially the config didn't had settings like 'daemon_name' and eventually that setting had more and more values.. But as the list grew I created the `new_gateway` dict, but left the previous code as it is.